connect-inject/endpoints_controller.go

```diff
@@ -378,49 +399,75 @@ func (r *EndpointsController) deregisterServiceOnAllAgents(ctx context.Context,
 		}),
 	}
 	if err := r.Client.List(ctx, &list, &listOptions); err != nil {
-		r.Log.Error(err, "failed to get agent pods from Kubernetes")
+		r.Log.Error(err, "failed to get Consul client agent pods")
 		return err
 	}
 
 	// On each agent, we need to get services matching "k8s-service-name" and "k8s-namespace" metadata.
 	for _, pod := range list.Items {
-		// Create client for this agent.
-		client, err := r.remoteConsulClient(pod.Status.PodIP)
-		if err != nil {
-			r.Log.Error(err, "failed to create a new Consul client", "address", pod.Status.PodIP)
-			return err
-		}
+		// Iterate through each namespace from the list of namespaces.
+		for _, ns := range r.consulNamespaceList() {
```

Comment:
Can the service instances be spread across consul namespaces for a single service? Why do we need to iterate through all namespaces in Consul? Would we be able to know what specific consul namespace to look at based on the k8snamespace input argument + the endpoints controller fields that configure namespaces? I'm probably just missing something here.

If we do need to iterate through every namespace can we add a comment about why?

Reply:
You are absolutely right!! I think I got a little carried away with the implementation of the cleanup controller where we dont have all the information in our hands!! Given we do have the k8s namespace, the consul namespace is in fact, deterministic. Gonna update the PR with this!!
cc @ishustava